ENIGMAsystem/SHELL/Graphics_Systems/General/GSvertex.cpp

```diff
@@ -0,0 +1,192 @@
+}
+
+void vertex_format_add_normal() {
+  enigma::vertexFormat->AddAttribute(vertex_type_float3, vertex_usage_normal);
+}
+
+void vertex_format_add_custom(int type, int usage) {
+  enigma::vertexFormat->AddAttribute(type, usage);
+}
+
+int vertex_format_end() {
+  int id = enigma::vertexFormats.size();
+  enigma::vertexFormats.push_back(enigma::vertexFormat);
+  return id;
+}
+
+bool vertex_format_exists(int id) {
+  return (id >= 0 && (unsigned)id < enigma::vertexFormats.size() && enigma::vertexFormats[id] != nullptr);
+}
+
+void vertex_format_delete(int id) {
+  delete enigma::vertexFormats[id];
+  enigma::vertexFormats[id] = nullptr;
+}
+
+int vertex_create_buffer() {
+  int id = enigma::vertexBuffers.size();
+  enigma::vertexBuffers.push_back(new enigma::VertexBuffer());
+  return id;
+}
+
+int vertex_create_buffer_ext(unsigned size) {
+  int id = enigma::vertexBuffers.size();
+  enigma::vertexBuffers.push_back(new enigma::VertexBuffer());
+  return id;
+}
+
+void vertex_delete_buffer(int buffer) {
+  enigma::graphics_delete_vertex_buffer_peer(buffer);
+  delete enigma::vertexBuffers[buffer];
+  enigma::vertexBuffers[buffer] = nullptr;
+}
+
+void vertex_begin(int buffer, int format) {
+  enigma::vertexBuffers[buffer]->vertices.clear();
+  enigma::vertexBuffers[buffer]->format = format;
+}
+
+void vertex_end(int buffer) {
+  enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+
+  // we can only flag the vertex buffer contents as dirty and needing an update
+  // if the vertex buffer hasn't been frozen, otherwise we just ignore it
+  if (vertexBuffer->frozen) return;
+  vertexBuffer->dirty = true;
+  vertexBuffer->number = vertexBuffer->vertices.size();
+}
+
+unsigned vertex_get_size(int buffer) {
+  return enigma::vertexBuffers[buffer]->number * sizeof(gs_scalar);
+}
+
+unsigned vertex_get_number(int buffer) {
+  const enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+  const enigma::VertexFormat* vertexFormat = enigma::vertexFormats[vertexBuffer->format];
+
+  return vertexBuffer->number / vertexFormat->stride;
+}
+
+void vertex_freeze(int buffer) {
+  // we can freeze the vertex buffer only if it isn't already frozen
+  // if it's not frozen, then we we'll freeze it when we do a dirty update
+  if (enigma::vertexBuffers[buffer]->frozen) return;
+  enigma::vertexBuffers[buffer]->frozen = true;
+  enigma::vertexBuffers[buffer]->dirty = true;
+}
+
+void vertex_submit(int buffer, int primitive) {
+  vertex_submit(buffer, primitive, 0, vertex_get_number(buffer));
+}
+
+void vertex_submit(int buffer, int primitive, int texture) {
+  texture_set(texture);
+  vertex_submit(buffer, primitive);
+}
+
+void vertex_submit(int buffer, int primitive, int texture, unsigned vertex_start, unsigned vertex_count) {
+  texture_set(texture);
+  vertex_submit(buffer, primitive, vertex_start, vertex_count);
+}
+
+void vertex_position(int buffer, gs_scalar x, gs_scalar y) {
+  enigma::vertexBuffers[buffer]->vertices.push_back(x);
```

Comment:
This code has no respect for vertex formats. It assumes everyone will always be writing 3D coordinates. That's really wasteful. Why do you have functions to define vertex formats if you just ignore them and send x,y,z triples for everything?

Reply:
Sorry, originally I was giving it just the x/y, the issue was `vertex_format_add_position` which should add float2, not float3. However, Direct3D9 is peculiar in that when you use the fixed-function pipeline it won't accept anything less than 3. So in that particular system I may have to go back into the vertices array and pack in a bunch of 0's. I have to investigate this, I totally forgot about it.